src/dbnode/persist/fs/merger.go

```diff
@@ -178,7 +178,8 @@ func (m *merger) Merge(
 			nsCtx.Schema,
 			encoderPool)
 	)
-	defer func() {
+	// This gets called later when we finish persist work.
+	cleanupFn := func() {
```

Comment:
Shouldn't this still get called if there's an error encountered?

Reply:
If we change the deferred close to not take a cleanup fn (since writing checkpoint file doesn't depend on these resources) we can revert this just to a straight up "defer" again.